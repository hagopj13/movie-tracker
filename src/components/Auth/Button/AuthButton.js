// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';

import DialogTypes from 'components/Dialog/types';
import authSelectors from 'store/auth/auth.selectors';
import authActions from 'store/auth/auth.actions';
import AuthActionTypes from 'store/auth/auth.types';
import loadingSelectors from 'store/api/loading/loading.selectors';
import dialogActions from 'store/ui/dialog/dialog.actions';

type Props = {
  isAuth: boolean,
  isLogoutLoading: boolean,
  onLogout: () => void,
  onShowLoginDialog: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    minWidth: 72,
    textTransform: 'none',
    fontSize: theme.typography.h6.fontSize,
    fontWeight: theme.typography.fontWeightRegular,
    '&$disabled': {
      color: theme.palette.grey[300],
    },
  },
  disabled: {},
  logout: {
    color: theme.palette.common.white,
  },
}));

const AuthButton = (props: Props) => {
  const { isAuth, isLogoutLoading, onLogout, onShowLoginDialog } = props;
  const classes = useStyles();

  return (
    <>
      {isAuth ? (
        <Button
          classes={{ root: classes.root, disabled: classes.disabled, text: classes.logout }}
          disabled={isLogoutLoading}
          onClick={onLogout}
        >
          Logout
        </Button>
      ) : (
        <Button className={classes.root} color="primary" onClick={onShowLoginDialog}>
          Login
        </Button>
      )}
    </>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
  isLogoutLoading: loadingSelectors.createIsLoadingSelector([AuthActionTypes.LOGOUT]),
});

const mapDispatchToProps = {
  onLogout: authActions.logout,
  onShowLoginDialog: () => dialogActions.showDialog(DialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(AuthButton);
