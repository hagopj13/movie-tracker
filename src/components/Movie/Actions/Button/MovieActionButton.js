// @flow
import React from 'react';
import type { Node } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';
import Fab from '@material-ui/core/Fab';
import { lightBlue } from '@material-ui/core/colors';

type Props = {
  children: Node,
  title: string,
  onClick: () => void,
};

const useStyles = makeStyles((theme) => ({
  tooltip: {
    fontSize: theme.typography.subtitle2.fontSize,
  },
  fab: {
    backgroundColor: lightBlue[800],
    '&:hover': {
      backgroundColor: lightBlue[900],
    },
  },
}));

const MovieActionButton = (props: Props, ref) => {
  const { children, title, onClick } = props;

  const classes = useStyles();

  return (
    <Tooltip classes={{ tooltip: classes.tooltip }} title={title}>
      <Fab className={classes.fab} ref={ref} size="medium" onClick={onClick}>
        {children}
      </Fab>
    </Tooltip>
  );
};

export default React.forwardRef<Props, Fab>(MovieActionButton);
