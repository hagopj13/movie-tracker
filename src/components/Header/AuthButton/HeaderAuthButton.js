// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';

import dialogTypes from 'config/dialogTypes';
import authSelectors from 'store/auth/auth.selectors';
import authActions from 'store/auth/auth.actions';
import AuthActionTypes from 'store/auth/auth.types';
import loadingSelectors from 'store/api/loading/loading.selectors';
import dialogActions from 'store/ui/dialog/dialog.actions';

type Props = {
  isAuth: boolean,
  isLogoutLoading: boolean,
  onLogout: () => void,
  onShowLoginDialog: () => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    minWidth: 72,
    textTransform: 'none',
    fontSize: theme.typography.h6.fontSize,
    '&$disabled': {
      color: theme.palette.grey[300],
    },
  },
  disabled: {},
  logoutText: {
    color: theme.palette.common.white,
  },
}));

export const HeaderAuthButton = (props: Props) => {
  const { isAuth, isLogoutLoading, onLogout, onShowLoginDialog } = props;
  const classes = useStyles();

  return isAuth ? (
    <Button
      classes={{ root: classes.root, disabled: classes.disabled, text: classes.logoutText }}
      disabled={isLogoutLoading}
      onClick={onLogout}
    >
      Logout
    </Button>
  ) : (
    <Button className={classes.root} color="primary" onClick={onShowLoginDialog}>
      Login
    </Button>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
  isLogoutLoading: loadingSelectors.createIsLoadingSelector([AuthActionTypes.LOGOUT]),
});

const mapDispatchToProps = {
  onLogout: authActions.logout,
  onShowLoginDialog: () => dialogActions.showDialog(dialogTypes.LOGIN),
};

export default connect(mapStateToProps, mapDispatchToProps)(HeaderAuthButton);
