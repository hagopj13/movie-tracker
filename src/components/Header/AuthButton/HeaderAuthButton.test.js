import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import Button from '@material-ui/core/Button';

import { HeaderAuthButton } from './HeaderAuthButton';

describe('HeaderAuthButton component', () => {
  let shallow;
  let wrapper;

  const mockLogout = jest.fn();
  const mockShowLoginDialog = jest.fn();

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('isAuth is true', () => {
    beforeEach(() => {
      const mockProps = {
        isAuth: true,
        isLogoutLoading: false,
        onLogout: mockLogout,
        onShowLoginDialog: mockShowLoginDialog,
      };
      wrapper = shallow(<HeaderAuthButton {...mockProps} />);
    });

    it('should render the HeaderAuthButton component correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the logout button', () => {
      expect(wrapper.find(Button).text()).toBe('Logout');
    });

    it('should call onLogout when the button is clicked', () => {
      wrapper.find(Button).simulate('click');
      expect(mockLogout).toHaveBeenCalledTimes(1);
    });

    it('should not disable the logout button when logout it not loading', () => {
      expect(wrapper.find(Button).prop('disabled')).toBe(false);
    });

    it('should disable the logout button when logout is loading', () => {
      const mockProps = {
        isAuth: true,
        isLogoutLoading: true,
        onLogout: mockLogout,
        onShowLoginDialog: mockShowLoginDialog,
      };
      const newWrapper = shallow(<HeaderAuthButton {...mockProps} />);
      expect(newWrapper.find(Button).prop('disabled')).toBe(true);
    });
  });

  describe('isAuth is false', () => {
    beforeEach(() => {
      const mockProps = {
        isAuth: false,
        isLogoutLoading: false,
        onLogout: mockLogout,
        onShowLoginDialog: mockShowLoginDialog,
      };
      wrapper = shallow(<HeaderAuthButton {...mockProps} />);
    });

    it('should render the HeaderAuthButton component correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the login button', () => {
      expect(wrapper.find(Button).text()).toBe('Login');
    });

    it('should call onShowLoginDialog when the button is clicked', () => {
      wrapper.find(Button).simulate('click');
      expect(mockShowLoginDialog).toHaveBeenCalledTimes(1);
    });
  });
});
