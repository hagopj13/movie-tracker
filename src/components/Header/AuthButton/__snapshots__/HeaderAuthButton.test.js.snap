// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderAuthButton component isAuth is false should render the HeaderAuthButton component correctly 1`] = `
<WithStyles(ForwardRef(Button))
  className="makeStyles-root-1"
  color="primary"
  onClick={[MockFunction]}
>
  Login
</WithStyles(ForwardRef(Button))>
`;

exports[`HeaderAuthButton component isAuth is true should render the HeaderAuthButton component correctly 1`] = `
<WithStyles(ForwardRef(Button))
  classes={
    Object {
      "disabled": "makeStyles-disabled-2",
      "root": "makeStyles-root-1",
      "text": "makeStyles-logoutText-3",
    }
  }
  disabled={false}
  onClick={[MockFunction]}
>
  Logout
</WithStyles(ForwardRef(Button))>
`;
